### PR TITLE
feat: added `fetchPriority`

### LINKED ISSUE #16989: "fetchPriority" for script tags
# Feature Request
Ability to set _fetchPriority_ attribute on script tags.

Webpack doesn't offer any way to set _fetchPriority_ attribute on script tags that are managed/added by webpack.

I propose adding `output.scriptFetchPriorty` option with values `boolean = false, string: 'high' | 'low' | 'auto'`
When !== false, it is used in LoadScriptRuntimeModule to add the attribute to the script tag.

## Relevant links
https://html.spec.whatwg.org/multipage/scripting.html#the-script-element
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#browser_compatibility

### PR DESCRIPTION
**Pull request by alexander-akait**
due to outdated fork I recreate a new 

fixes #16991
fixes #16989

### <samp>🤖 Generated by Copilot at 4b6eed7</samp>

This pull request adds a new feature to webpack that allows setting the fetch priority of dynamic imports, either globally or individually. This feature can improve the loading performance of async chunks by using the `fetchPriority` attribute on script tags. The pull request modifies several files in the lib, schemas and test folders to implement, validate and test this feature. It also updates the `cspell.json` file to add a new word related to this feature.

### <samp>🤖 Generated by Copilot at 4b6eed7</samp>

*  Add a new feature to allow setting the fetch priority for dynamic imports using the `fetchPriority` attribute on script tags ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-77519f787fe7e051a14c588f101d61e4d8f0d3b01bf8e16cb65d3838243d4e68R84), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-3d8c1094b9972f01c64b52217e93034969d92c01fae59dd8fb841ac97a287950R31), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-c0078e4713a0e7a24dbc56d6dfda3cf139376325fc65ea70a8389d4d30b27f04R493), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-58a908cfddf2545aff58acb9a48335b7d76d68bbc21f5428cc71f9cbee10b50aR222-R242), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-391177474567b080c642e311989bb34a9a82be71c4d65e15fc2ca97e73caaf5aL50-R54), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-391177474567b080c642e311989bb34a9a82be71c4d65e15fc2ca97e73caaf5aR64-R71), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-391177474567b080c642e311989bb34a9a82be71c4d65e15fc2ca97e73caaf5aR156-R172), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-b732193070dd45bf63e923b349ded24a6467b4a06906e7e4dd6ee3a4ed076853R1-R41), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-cdcce88f2f19b714e7763e61ac874b2fcf63ee41649a9179db8b388c0ee9b688R9), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-cdcce88f2f19b714e7763e61ac874b2fcf63ee41649a9179db8b388c0ee9b688L46-R47), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-cdcce88f2f19b714e7763e61ac874b2fcf63ee41649a9179db8b388c0ee9b688R91-R119), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-848f7b6e954a1b22b9b6541b6ae06f778f4fd0b4b95c44ee86c2b1c43f7f0b90L47-R52), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-848f7b6e954a1b22b9b6541b6ae06f778f4fd0b4b95c44ee86c2b1c43f7f0b90R86-R94), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-848f7b6e954a1b22b9b6541b6ae06f778f4fd0b4b95c44ee86c2b1c43f7f0b90L108-R170), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-99b22d1932803b7795bbe843d01cdf63d32e52a247610bfd9cffb95170227cc6R111-R113), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-99b22d1932803b7795bbe843d01cdf63d32e52a247610bfd9cffb95170227cc6R172-R174), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-99b22d1932803b7795bbe843d01cdf63d32e52a247610bfd9cffb95170227cc6L193-R203), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-0cdb656a2796a80418de2964257360de642bd5a770102d52366890a65aa912a0R1623-R1633), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-0f5f45f99dc18408e253cdcdbaae82c9a756182a0b01e3fbb768c21856a60e70R183-R189))
  * Add a new parser option `dynamicImportFetchPriority` to specify the global fetch priority for dynamic imports in `lib/config/defaults.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-c0078e4713a0e7a24dbc56d6dfda3cf139376325fc65ea70a8389d4d30b27f04R493))
  * Add a new property `fetchPriority` to the `ChunkGroup` class to store the fetch priority option for a chunk group in `lib/ChunkGroup.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-3d8c1094b9972f01c64b52217e93034969d92c01fae59dd8fb841ac97a287950R31))
  * Add a new case for the `fetchPriority` property in the `ImportMetaContextDependencyParserPlugin` class to parse the `import.meta` object for dynamic imports in `lib/dependencies/ImportMetaContextDependencyParserPlugin.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-58a908cfddf2545aff58acb9a48335b7d76d68bbc21f5428cc71f9cbee10b50aR222-R242))
  * Add a new condition to the `ImportParserPlugin` class to check the `dynamicImportFetchPriority` option and set the `groupOptions.fetchPriority` accordingly in `lib/dependencies/ImportParserPlugin.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-391177474567b080c642e311989bb34a9a82be71c4d65e15fc2ca97e73caaf5aR64-R71))
  * Add a new condition to the `ImportParserPlugin` class to check the `webpackFetchPriority` property in the `importOptions` object and set the `groupOptions.fetchPriority` accordingly or add a warning in `lib/dependencies/ImportParserPlugin.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-391177474567b080c642e311989bb34a9a82be71c4d65e15fc2ca97e73caaf5aR156-R172))
  * Add a new class `ChunkFetchPriorityFunctionRuntimeModule` to generate the runtime code for the `getFetchPriority` function that returns the fetch priority for a given chunk id in `lib/prefetch/ChunkFetchPriorityFunctionRuntimeModule.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-b732193070dd45bf63e923b349ded24a6467b4a06906e7e4dd6ee3a4ed076853R1-R41))
  * Add a new hook to the `ChunkPrefetchPreloadPlugin` class to iterate over the async chunks and collect the fetch priorities and add a new `ChunkFetchPriorityFunctionRuntimeModule` and the `getFetchPriority` requirement if needed in `lib/prefetch/ChunkPrefetchPreloadPlugin.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-cdcce88f2f19b714e7763e61ac874b2fcf63ee41649a9179db8b388c0ee9b688R91-R119))
  * Add a new parameter `hasGetFetchPriority` to the `LoadScriptRuntimeModule` class to determine if the `loadScript` function should use the `fetchPriority` attribute or not in `lib/runtime/LoadScriptRuntimeModule.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-848f7b6e954a1b22b9b6541b6ae06f778f4fd0b4b95c44ee86c2b1c43f7f0b90L47-R52))
  * Add a new condition to the `LoadScriptRuntimeModule` class to generate the code to set the `fetchPriority` attribute on the script tag using the `getFetchPriority` function and the chunk id in `lib/runtime/LoadScriptRuntimeModule.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-848f7b6e954a1b22b9b6541b6ae06f778f4fd0b4b95c44ee86c2b1c43f7f0b90R86-R94))
  * Add a new argument `fetchPriority` to the `loadScript` function signature and pass it to the `createScript` function call in `lib/runtime/LoadScriptRuntimeModule.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-848f7b6e954a1b22b9b6541b6ae06f778f4fd0b4b95c44ee86c2b1c43f7f0b90L108-R170))
  * Add a new runtime global function `getFetchPriority` to return the fetch priority for a given chunk id in `lib/RuntimeGlobals.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-c3d58ece2e36cf9e319d260abb918fbfa8c8f0948c95d3f39b29907ea22ed30fR192-R197))
  * Add a new condition to the `RuntimePlugin` class to check the `getFetchPriority` requirement and pass it as an argument to the `LoadScriptRuntimeModule` constructor in `lib/RuntimePlugin.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-8ca568ebcc2831549e2d258ee4a9aeeb1b84ca32394824ac94c3de622c1f6606L376-R382))
  * Add a new condition to the `JsonpChunkLoadingRuntimeModule` class to check the `getFetchPriority` requirement and store it in a local variable `withFetchPriority` in `lib/web/JsonpChunkLoadingRuntimeModule.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-99b22d1932803b7795bbe843d01cdf63d32e52a247610bfd9cffb95170227cc6R111-R113))
  * Add a new condition to the `JsonpChunkLoadingRuntimeModule` class to generate the code to declare and assign the `fetchPriority` variable using the `getFetchPriority` function and the chunk id in `lib/web/JsonpChunkLoadingRuntimeModule.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-99b22d1932803b7795bbe843d01cdf63d32e52a247610bfd9cffb95170227cc6R172-R174))
  * Add a new argument `fetchPriority` to the `loadScript` function call and pass it if the `withFetchPriority` variable is true in `lib/web/JsonpChunkLoadingRuntimeModule.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-99b22d1932803b7795bbe843d01cdf63d32e52a247610bfd9cffb95170227cc6L193-R203))
  * Add a new schema definition for the `dynamicImportFetchPriority` option to validate the webpack configuration in `schemas/WebpackOptions.json` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-0cdb656a2796a80418de2964257360de642bd5a770102d52366890a65aa912a0R1623-R1633))
  * Add a new test case to test the inline `webpackFetchPriority` option for dynamic imports in `test/cases/chunks/inline-options/index.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-0f5f45f99dc18408e253cdcdbaae82c9a756182a0b01e3fbb768c21856a60e70R183-R189))
    * Add new files `a.js`, `b.js` and `c.js` to test the inline `webpackFetchPriority` option for dynamic imports in `test/cases/chunks/inline-options/dir14/` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-6dfbe1b3872acbd9cac362cb41cd709462ad7cd1ab9b9a742eab3e032713ed39R1), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-e7425bea3d5a074379ef498820dfe36d734cfd6e95ca607c44f797ad304ca732R1), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-5d1748a4a58d342773225732c41e1a08f850ef09882f841d073d197949efac23R1))
* Swap the order of the words "IIFE" and "IIFE's" and "webpack" and "webpack's" in the `cspell.json` file to keep the words in alphabetical order ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-77519f787fe7e051a14c588f101d61e4d8f0d3b01bf8e16cb65d3838243d4e68L107-R109), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-77519f787fe7e051a14c588f101d61e4d8f0d3b01bf8e16cb65d3838243d4e68L277-R279))
* Remove the second argument of false from the call to the `getChildIdsByOrdersMap` method of the chunk object in the `ChunkPrefetchPreloadPlugin` class, which is not needed since the fetch priority is optional ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-cdcce88f2f19b714e7763e61ac874b2fcf63ee41649a9179db8b388c0ee9b688L46-R47))
* Destructure the `dynamicImportFetchPriority` option from the options object and store it in a local variable in the `ImportParserPlugin` class, which is a minor refactoring ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-391177474567b080c642e311989bb34a9a82be71c4d65e15fc2ca97e73caaf5aL50-R54))
* Add a new file `a.js` to test the global `dynamicImportFetchPriority` option for dynamic imports in `test/configCases/web/fetch-priority-2/` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-b84f1ca6a37989b19e46ca52ffee07a7d7335b849b43e9b479aa4bdea297b6ceR1-R4))
  * Add new files `b.js` and `c.js` to test the global `dynamicImportFetchPriority` option for dynamic imports in `test/configCases/web/fetch-priority-2/` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-c541f1d45b7286525313c6973598090bffac70527ef29020d3e4cdcc93d8254eR1), [link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-3593266d406f2108031489020c94e2edc93b5f730b8e97ceed5d2926b2745b53R1))
* Add a new default value for the `dynamicImportFetchPriority` option to test the default values of the webpack configuration in `test/Defaults.unittest.js` ([link](https://github.com/webpack/webpack/pull/17249/files?diff=unified&w=0#diff-2ca9058acdbb297c64353f1f82fdd9828c21911ce176861026c6a0bea93fb5c2R218))

### CODE REVIEW

**Comment by alexander-akait** [comment]
@aboktor @Zlatkovsky please take a look, fixed:
- no boolean value anymore (except `false`, only for configuration, it means chunks groups don't have fetchpriorty attribute by default)
- do not override and use the first found value, for others groups it no longer matters because after loading chunk we will not load it anymore

**Comment by TheLarkInn** [comment]
@aboktor Hey would you like to review this and then inquire if there is anything that doesn't seem clear. We'll discuss and then we will land this in next weeks release.

**Comment by TheLarkInn** [comment]
Also @Zlatkovsky would you mind re-reviewing as well. Just want to make sure we are all aligned on the feature/goals/api surface and that this solves your scenario.

**Comment by aboktor** [comment]
Took a look through this. I want to validate my understanding. Here is how I understand things in their current state:
1. fetchPriority for a chunk is computed at bundle time (my initial PR was computing fetchPriority for a chunk at runtime, based on which async entrypoint requested it). I want to make sure that it is a conscious decision that we want fetchPriority to be decided at bundle time.
2. fetchPriority will be taken from the async entrypoint with the highest prefetchOrder if more than one exists.

Questions:
Is prefetchOrder computed based on the dependency graph and thus makes sense to use here to decide on fetchPriority?

Comments:
I still think that #2 above is a bit counter intuitive. One thing we could do here is to take the highest fetchPriority from all async entrypoints that require a particular chunk ( undefined < "auto" < "low" < "high" ).

@alexander-akait

**Review by Zlatkovsky — APPROVED** [review]
Looks good to me. For what it's worth, I know that @aboktor tried setting fetch-priority (just via a quick patch) in an app he was working on, and did not see a statistically-significant end-to-end difference in how quickly that specific app loaded. I have an opportunity to try it out as well in a different context, but won't know if it helped or not for another few weeks.

It still seems like this should only be a good thing; but just an FYI that we don't yet have an empirical-evidence case of where setting the fetch priority helps.

**Comment by alexander-akait** [comment]
> Questions:
> Is prefetchOrder computed based on the dependency graph and thus makes sense to use here to decide on fetchPriority?

You can have `prefetchOrder` (and not have, in this case it is just `0`), so it will be used only when `prefetchOrder` exists, to be compatibility with with prefetch 

> I still think that https://github.com/webpack/webpack/pull/2 above is a bit counter intuitive. One thing we could do here is to take the highest fetchPriority from all async entrypoints that require a particular chunk ( undefined < "auto" < "low" < "high" ).

Yeah, let's add tests too and give me time to think about it (today), we should not increase our runtime when developer doesn't use it, but I see what do you mean

We need to ship it for the next release :smile:

**Comment by alexander-akait** [comment]
@aboktor @Zlatkovsky 

I refactor again and added tests, because I agree when you if something `if (value) { await import(...) } else { await import(...) }` you will have magic and can't predicate what will be used, but now we generate code only when it requires, so your runtime will be the same if you don't use `fetchPriority`, now should be no problems (I hope :smile: )

There is intresting thing - we use `import(...)` for es2020 and web target, and because `import(...)` doesn't have the such ability, you can't use this feature, I would say it's not consistent between the specs (ES modules and HTML)

File: test/__snapshots__/StatsTestCases.basictest.js.snap [hunk 1]
```diff
@@ -3,12 +3,12 @@
 exports[`StatsTestCases should print correct stats for aggressive-splitting-entry 1`] = `
 "fitting:
   PublicPath: auto
-  asset fitting-42703728faa2ac5f1783.js 16.1 KiB [emitted] [immutable]
+  asset fitting-801f5840b67f0f13ee03.js 16.2 KiB [emitted] [immutable]
   asset fitting-50595d23e8f97d7ccd2a.js 1.9 KiB [emitted] [immutable]
   asset fitting-5bc77880fdc9e2bf09ee.js 1.9 KiB [emitted] [immutable]
   asset fitting-72afdc913f6cf884b457.js 1.08 KiB [emitted] [immutable]
-  Entrypoint main 19.9 KiB = fitting-50595d23e8f97d7ccd2a.js 1.9 KiB fitting-5bc77880fdc9e2bf09ee.js 1.9 KiB fitting-42703728faa2ac5f1783.js 16.1 KiB
```
Comment (alexander-akait) [inline test/__snapshots__/StatsTestCases.basictest.js.snap:10]: We have just an extra newline due concatenation logic in runtime template, not affected on production output

**Review by TheLarkInn — APPROVED** [review]

**Comment by TheLarkInn** [comment]
@aboktor Thank you for your first contribution to webpack! 🎉

**Pull request merged**